File: 백준/Gold/2155. 삼각형의 최단 경로/삼각형의 최단 경로.py
# 삼각형의 최단 경로
import sys
import math
input = sys.stdin.readline

start, end = map(int,input().split())

if start > end:
    start, end = end, start

startLevel = math.ceil(start**(0.5))
endLevel = math.ceil(end**(0.5))    
num = start # 레벨 이동이 끝난 후 제일 왼쪽의 숫자
pos = 0 # 레벨 이동이 끝난 후 가능한 숫자의 갯수

if start % 2 == startLevel % 2:
    move = - 1
    num += 1
else:
    move = 0
    pos += 1

move += (endLevel - startLevel) * 2
levelMove = startLevel * 2 - 1

while startLevel < endLevel:
    num += levelMove
    levelMove += 2
    startLevel += 1
    pos += 1

minn = abs(num - end)
for i in range(1, pos):
    minn = min(minn, abs((i*2 + num) - end))

print(move + minn)

File: 백준/Gold/2155. 삼각형의 최단 경로/삼각형의 최단 경로.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int start = Integer.parseInt(st.nextToken());
        int end = Integer.parseInt(st.nextToken());

        if (start > end) {
            int tmp = start;
            start = end;
            end = tmp;
        }

        int startLevel = (int) Math.ceil(Math.sqrt(start));
        int endLevel = (int) Math.ceil(Math.sqrt(end));
        int num = start; // 레벨 이동이 끝난 후 제일 왼쪽의 숫자
        int pos = 0; // 레벨 이동이 끝난 후 가능한 숫자의 갯수
        int move = 0;

        if (start % 2 == startLevel % 2) {
            move = -1;
            num += 1;
        } else {
            pos += 1;
        }

        move += (endLevel - startLevel) * 2;
        int levelMove = startLevel*2 - 1;

        while (startLevel < endLevel) {
            num += levelMove;
            levelMove += 2;
            startLevel += 1;
            pos += 1;
        }

        int min = Math.abs(num - end);

        for (int i = 1; i < pos; i++) {
            min = Math.min(min, Math.abs((i*2 + num) - end));
        }

        System.out.println(move + min);
    }
}
